### PR TITLE
Rfc4585

### PR DESCRIPTION
**Pull request by funman**
Comes with the apps to test it:

 ./udpmulticat_fb @127.0.0.1:1234 192.168.0.15:1234 200

Relays RTP on @:1234 to remote, with 200ms buffer

 ./udpmulticat -q @192.168.0.15:1234 127.0.0.1:7777 200

Receives and repairs RTP then send it to 7777, same 200ms buffer

To add some difficulties on sender:
sudo tc qdisc add dev enp0s31f6 root netem
sudo tc qdisc change dev enp0s31f6 root netem delay 100ms
sudo tc qdisc change dev enp0s31f6 root netem loss 2%

### CODE REVIEW

File: lib/upipe-modules/upipe_rtp_feedback.c [hunk 1]
```diff
@@ -0,0 +1,1034 @@
...
+        case UPIPE_SET_OUTPUT: {
+            struct upipe *output = va_arg(args, struct upipe *);
+            return upipe_rtpfb_set_output(upipe, output);
+        }
+        case UPIPE_SET_OPTION: {
+            const char *k = va_arg(args, const char *);
+            const char *v = va_arg(args, const char *);
+            return upipe_rtpfb_set_option(upipe, k, v);
+        }
+        case UPIPE_CONTROL_LOCAL:
```
Comment (cmassiot) [inline lib/upipe-modules/upipe_rtp_feedback.c]: Generally the idea is to have a pipe-specific control, and an inline primitive in the header file.

File: examples/udpmulticat.c [hunk 2]
```diff
@@ -75,34 +77,125 @@
...
+static int udp_fd = -1;
+
+static struct sockaddr_storage addr;
+static socklen_t addr_len;
 
 static void usage(const char *argv0) {
-    fprintf(stdout, "Usage: %s [-d] [-r <rotate>] [-O <rotate offset>] <udp source> <dest dir/prefix> [<suffix>]\n", argv0);
+    fprintf(stdout, "Usage: %s [-d]  <udp source> <dest dir/prefix> <latency>", argv0);
     fprintf(stdout, "   -d: force debug log level\n");
-    fprintf(stdout, "   -r: rotate interval in 27MHz unit\n");
```
Comment (cmassiot) [inline examples/udpmulticat.c]: Why do you remove features from the examples? If it's too hard to have both use cases in the same example, create a new one.

File: examples/udpmulticat.c [hunk 3]
```diff
@@ -113,158 +206,184 @@ static int catch(struct uprobe *uprobe, struct upipe *upipe,
...
+    upipe_attach_uclock(upipe_udpsrc);
+
+    /* send to udp */
+
+    upipe_set_output(upipe_rtpfb, upipe_udp_sink);
+    upipe_release(upipe_udp_sink);
+
     /* fire loop ! */
-    upump_mgr_run(upump_mgr, NULL);
+    ev_loop(loop, 0);
```
Comment (cmassiot) [inline examples/udpmulticat.c]: upump_mgr_run is now the preferred way to start the upump manager.

File: examples/udpmulticat.c [hunk 4]
```diff
@@ -113,158 +206,184 @@ static int catch(struct uprobe *uprobe, struct upipe *upipe,
...
+    struct ev_loop *loop = ev_default_loop(0);
     struct umem_mgr *umem_mgr = umem_alloc_mgr_alloc();
     struct udict_mgr *udict_mgr = udict_inline_mgr_alloc(UDICT_POOL_DEPTH,
                                                          umem_mgr, -1, -1);
     struct uref_mgr *uref_mgr = uref_std_mgr_alloc(UREF_POOL_DEPTH, udict_mgr,
                                                    0);
-    struct upump_mgr *upump_mgr = upump_ev_mgr_alloc_default(UPUMP_POOL,
-            UPUMP_BLOCKER_POOL);
-    struct uclock *uclock = uclock_std_alloc(UCLOCK_FLAG_REALTIME);
+    struct upump_mgr *upump_mgr = upump_ev_mgr_alloc(loop, UPUMP_POOL,
```
Comment (cmassiot) [inline examples/udpmulticat.c]: upump_ev_mgr_alloc_default avoids having to create an ev_loop.

Comment (funman) [reply examples/udpmulticat.c]: I didn't update after the upump changes

File: examples/udpmulticat.c (hunk 2 above)
Comment (cmassiot) [reply examples/udpmulticat.c]: IMO it's always interesting to have an example

Comment (funman) [reply examples/udpmulticat.c]: @kierank Which license for the examples?

File: examples/arq_rx.c [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+
+/** definition of our uprobe */
+static int catch_udp(struct uprobe *uprobe, struct upipe *upipe,
+                 int event, va_list args)
+{
+    switch (event) {
+    case UPROBE_UDPSRC_NEW_PEER: {
+        int sig = va_arg(args, int);
+        if (sig != UPIPE_UDPSRC_SIGNATURE)
+            return UBASE_ERR_INVALID;
```
Comment (cmassiot) [inline examples/arq_rx.c]: Theoritically you should pass it to the next probe.

**Pull request closed without merging**